### DISCUSSION TITLE
wagmi v2 analogue of `connector.watchAsset`?

### DISCUSSION THREAD

**Discussion by firnprotocol** [wevm/wagmi#3345] [General]
indeed, it seems that `watchAsset` has been removed from the `connector` datatype.

what is the correct way to "request wallet add token" in wagmi v2? thanks in advance.

sketch of current code:
```javascript
export function useAddToWallet() {
  const { address, connector, chain } = useAccount();

  const [adding, setAdding] = useState(false); // need to keep this around, i.e., can't _just_ use `locked`,
  const [added, setAdded] = useState(false);

  async function onClick(event) {
    setAdding(true);
    try {
      const result = await connector.watchAsset({ // doesn't exist anymore, at least for injected metamask...
        address: ADDRESSES["Ethereum"]["ERC20"],
        symbol: "FIRN",
        decimals: 18,
        image: `${window.location.href}android-chrome-256x256.png`,
      });
      if (result) setAdded(true);
    } catch (error) {
      console.error(error);
      if (error.code === 4001)
        toast.error("You declined the token addition prompt.");
      else if (error.message === "connector.watchAsset is not a function")
        toast.error("Your currently selected wallet doesn't support this action.");
      else
        toast.error("An unknown error occurred.");
    } finally {
      setAdding(false);
    }
  }

  return { onClick, adding, added };
}
```

**Comment by firnprotocol** [comment]
update: trying this as well:
```javascript
      const walletClient = getWalletClient(config);
      const result = await walletClient.watchAsset({
        type: 'ERC20',
        options: {
          address: ADDRESSES["Ethereum"]["ERC20"],
          symbol: "FIRN", // A ticker symbol or shorthand, up to 5 chars.
          decimals: 18, // The number of decimals in the token
          image: `${window.location.href}android-chrome-256x256.png`,
        }
      });
```
based on [this](https://rc.viem.sh/docs/actions/wallet/watchAsset.html) it seems that this should work, but does not:
```javascript
TypeError: walletClient.watchAsset is not a function
    at _callee$ (useAddToWallet.js:30:41)
    at tryCatch (regeneratorRuntime.js:44:1)
    at Generator.<anonymous> (regeneratorRuntime.js:125:1)
    at Generator.next (regeneratorRuntime.js:69:1)
    at asyncGeneratorStep (asyncToGenerator.js:3:1)
    at _next (asyncToGenerator.js:22:1)
    at asyncToGenerator.js:27:1
    at new Promise (<anonymous>)
    at asyncToGenerator.js:19:1
    at _onClick (useAddToWallet.js:53:4)
```

**Comment by firnprotocol** [comment]
hi all, i wonder if i could bump this please? is there a way to "watch asset" in wagmi 2? thanks in advance.

**Reply by tmm** [reply]
Since there is [already a Viem Action](https://viem.sh/docs/actions/wallet/watchAsset), you can either [drop down to Viem](https://wagmi.sh/react/guides/viem#leveraging-viem-actions) or we will happily accept a contribution for this. Just need to add the core action, query helper, and hook. Here's [a good recent example](https://github.com/wevm/wagmi/pull/3461) of one of those pull requests from another contributor.

**Reply by firnprotocol** [reply]
hmm, but even after setting `const { data: walletClient } = useConnectorClient();`, I am still finding that the property `watchAsset` is missing from the resulting `walletClient`, even on (injected) MetaMask... what am i missing?!

**Reply by tmm** [reply]
Take another look at the example I linked: https://wagmi.sh/react/guides/viem#leveraging-viem-actions

**Reply by firnprotocol** [reply]
i'm an idiot. thanks.

will look into contributing this later tonight

**Reply by firnprotocol** [reply]
btw, FWIW, i had been confused by this page: https://viem.sh/docs/actions/wallet/watchAsset I'm assuming that that page is inaccurate?

**Reply by tmm** [reply]
What is confusing about it?

**Reply by firnprotocol** [reply]
the above page seems to suggest that each object of type "wallet client" should have a member `watchAsset`; whereas, in the page _you_ linked, `watchAsset` is actually a standalone, top-level function in `viem/actions`, which is _passed in_ a `walletClient`, together with the token information....right?!

**Reply by tmm** [reply]
Correct, actions can generally be used in two ways: [tree-shakable](https://viem.sh/docs/clients/custom#tree-shaking) (recommended) and via client decorators.

**Reply by firnprotocol** [reply]
ok, but in my setting, the latter approach (i.e., top-level `watchAsset(walletClient, { ... })`) "worked", while the former approach (`walletClient.watchAsset({ ... })`) did not. why the difference?

**Reply by tmm** [reply]
Connectors return a Connector Client by default instead of a Wallet Client. Wallet Client has all the actions exposed off of it and is not tree-shakable. As a result, it tends to blow up bundle size and have a lot of unused code associated with it.

If you want to convert a Connector Client to a Wallet Client, you can use Viem's [`.extend`](https://viem.sh/docs/clients/custom#extending-with-actions-or-configuration) API, but I wouldn't recommend it.

**Comment by einaralex** [comment]
Stumbled onto this discussion while trying to solve this, this is my solution.


```js
import { EIP1193Provider } from 'viem';
import { useAccount } from 'wagmi';

export const useWatchAsset = () => {
  const { isConnected, connector } = useAccount();

  if (!isConnected || !connector) {
    return { watchAsset: null };
  }

  return {
    watchAsset: async (token) => {
      try {
        const provider = await connector.getProvider();

        await (provider as EIP1193Provider)?.request({
          // See details: https://docs.metamask.io/wallet/reference/wallet_watchasset/
          method: 'wallet_watchAsset',
          params: {
            type: 'ERC20',
            options: token,
          },
        });
      } catch (error) {
        console.error('Failed to watch asset:', error);
      }
    },
  };
};
```

Usage
```js
const { watchAsset } = useWatchAsset();


if (watchAsset) {
  watchAsset({ address: tokenAddress })
}
   
```
